include/onnxruntime/core/framework/sparse_tensor.h

```diff
@@ -1,72 +1,434 @@
-  ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(SparseTensor);
+  ORT_DISALLOW_COPY_AND_ASSIGNMENT(SparseTensor);
 
-  // Returns the number of entries in the values tensor (aka "NNZ" or "number of nonzero values")
+  /// <summary>
+  // Returns the number of non-zero values (aka "NNZ")
+  // For block sparse formats this may include some zeros in the blocks
+  // are considered non-zero.
+  /// </summary>
+  /// <returns>nnz</returns>
   size_t NumValues() const { return static_cast<size_t>(values_.Shape().Size()); }
 
-  const Tensor& Values() const {
+  /// <summary>
+  /// Read only accessor to non-zero values
+  /// </summary>
+  /// <returns></returns>
+  const Tensor& Values() const noexcept {
     return values_;
   }
 
-  const Tensor& Indices() const {
-    return indices_;
+  SparseTensor(SparseTensor&& o) noexcept;
+  SparseTensor& operator=(SparseTensor&& o) noexcept;
+
+  /// <summary>
+  /// Returns SparseFormat that the instance currently holds
+  /// if the value returned in kUndefined, the instance is not populated
+  /// </summary>
+  /// <returns>format enum</returns>
+  SparseFormat Format() const noexcept {
+    return format_;
+  }
+
+  /// <summary>
+  /// Returns a would be dense_shape
+  /// </summary>
+  /// <returns>reference to dense_shape</returns>
+  const TensorShape& DenseShape() const noexcept {
+    return dense_shape_;
+  }
+
+  /// <summary>
+  /// Calculates and returns how much this fully initialized SparseTensor data (would)
+  /// occupy in a contiguous allocation block, or, in fact, occupies if it owns the buffer.
+  /// </summary>
+  /// <returns>required allocation size</returns>
+  int64_t RequiredAllocationSize() const noexcept;
+
+  /// <summary>
+  /// Returns Tensor element type enum.
+  /// Useful for type dispatching
+  /// </summary>
+  /// <returns></returns>
+  int32_t GetElementType() const {
+    return ml_data_type_->GetDataType();
   }
 
-  const TensorShape& Shape() const {
-    return shape_;
+  /// <summary>
+  /// Return Element MLDataType
+  /// </summary>
+  /// <returns></returns>
+  MLDataType DataType() const noexcept {
+    return ml_data_type_;
   }
 
-  Tensor& MutableValues() {
-    return values_;
+  /// <summary>
+  /// Test for string type
+  /// </summary>
+  /// <returns>true if tensor values are strings</returns>
+  bool IsDataTypeString() const {
+    return utils::IsPrimitiveDataType<std::string>(ml_data_type_);
   }
 
-  Tensor& MutableIndices() {
-    return indices_;
+  /// <summary>
+  /// Checks if the Tensor contains data type T
+  /// </summary>
+  /// <typeparam name="T"></typeparam>
+  /// <returns>true if tensor contains data of type T</returns>
+  template <class T>
+  bool IsDataType() const {
+    return utils::IsPrimitiveDataType<T>(ml_data_type_);
   }
 
-  //TensorShape& MutableShape() {
-  //  return shape_;
-  //}
+  const OrtMemoryInfo& Location() const noexcept { return location_; }
+
+  /// <summary>
+  /// Read only access to Coo indices
```

Comment:
I am a bit mystified why we have a read-only access to just the indices, instead of indices+values?

Reply:
Values read only access is provided as a part of class interface. Views are format specific, so they expose only format specific parts.